planner/core/rule_semi_join_rewrite.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"context"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/expression/aggregation"
+	"github.com/pingcap/tidb/parser/ast"
+)
+
+type semiJoinRewriter struct {
+}
+
+func (smj *semiJoinRewriter) optimize(_ context.Context, p LogicalPlan, opt *logicalOptimizeOp) (LogicalPlan, error) {
+	return smj.recursivePlan(p)
+}
+
+func (smj *semiJoinRewriter) name() string {
+	return "semi_join_rewrite"
+}
+
+func (smj *semiJoinRewriter) recursivePlan(p LogicalPlan) (LogicalPlan, error) {
+	newChildren := make([]LogicalPlan, 0, len(p.Children()))
+	for _, child := range p.Children() {
+		newChild, err := smj.recursivePlan(child)
+		if err != nil {
+			return nil, err
+		}
+		newChildren = append(newChildren, newChild)
+	}
+	p.SetChildren(newChildren...)
+	join, ok := p.(*LogicalJoin)
+	// If it's not a join, or not a (outer) semi join. We just return it since no optimization is needed.
+	// Actually the check of the preferRewriteSemiJoin is a superset of checking the join type. We remain them for a better understanding.
+	if !ok || !(join.JoinType == SemiJoin || join.JoinType == LeftOuterSemiJoin) || (join.preferJoinType&preferRewriteSemiJoin == 0) {
+		return p, nil
+	}
+
+	if join.JoinType == LeftOuterSemiJoin {
+		p.SCtx().GetSessionVars().StmtCtx.AppendWarning(ErrInternal.GenWithStack("The EXISTS/IN subquery is too complicated to apply the SEMI_JOIN_REWRITE hint."))
```

Comment:
I think it's better to make the reason clear, like "SEMI_JOIN_REWRITE() is inapplicable for LeftOuterSemiJoin".

Reply:
"Hint SEMI_JOIN_REWRITE is not applicable to LeftOuterSemiJoin. Please remove this hint"